Resolve #614 - a few bug fixes in AC train-test split

igiloh opened the pull request:
1. Solved two bugs in the way the train-validation split was used in the Activity Classifier (causing `create()` to crash when `validation_set='auto'`).
2. Improved the `random_split_by_session() ` mechanism. The current implementation might have broken line order - which is not allowed for the AC. The new mechanism is optimized for the typical case of 100-1000 training sessions.
3. Added explicit unit test that covers the auto validation set in all relevant cases.

nickjong commented on src/unity/python/turicreate/toolkits/activity_classifier/util.py:
```diff
@@ -68,7 +70,43 @@ def random_split_by_session(dataset, session_id, fraction=0.9, seed=None):
+        not_chosen_filter = data[session_id].is_in(not_chosen['session'])
+
+        train = data[1 - not_chosen_filter]
+        valid = data[not_chosen_filter]
```
All of this logic makes sense, but it's frustrating, since random_split works by generating a random Boolean value for each row and then performing a filter internally, before returning the two resulting SArrays. We're having to re-implement so much of this operation, and the is_in case requires a linear search through `not_chosen` for each row, and the filter case requires sorting.

Could we somehow generate a random boolean for each row that yields the same result for each session ID? In principle, all it would take is a suitable hash function:

```
chosen_filter = data[session_id].apply(lambda x: hash_fn(x + seed) < fraction * MAX_HASH_FN_OUTPUT)
train = data[chosen_filter]
valid = data[1 - chosen_filter]
```

The total operation should then be linear, at face value. All that is required is a suitable hash function. For `hash_fn(x)` it may even suffice to re-initialize a RNG with `x` as the seed, perform one draw from uniform(0,1) and compare against `fraction`.

If there's something I'm missing, or if Python doesn't really make it easy enough to do the above, I could always investigate this approach as a followup someday. How long does this approach take in practice?

igiloh replied:
Hmm - interesting. This sounds like a good idea.  
My only concern is that it only guarantees that the sessions are actually divided according to the divided fraction in expectancy. For a small number of sessions (which is the typical case) - the actual fraction might be quite different.  
I'm not sure how accurate our current `random_split` mechanism is (which we use to divide the sessions). Maybe @ylow has some more information.

I have a bolt task where I've compared the `is_in` and the `filter` implementations for various dataset sizes (number of sessions). I'll run it again using your idea - and compare run time and fraction accuracy.

nickjong replied:
Ah, good point about the variance in the split. Looking at the implementation: [unity_sframe::random_split](https://github.com/apple/turicreate/blob/e43e2133d725cf8cbdb3d92f8d4d46facb58a3aa/src/unity/lib/unity_sframe.cpp#L1071) calls into [unity_sarray::make_uniform_boolean_array](https://github.com/apple/turicreate/blob/e43e2133d725cf8cbdb3d92f8d4d46facb58a3aa/src/unity/lib/unity_sarray.cpp#L1690), which uses essentially the idea above.

Interestingly, the C++ version has an `exact` parameter that causes it split as close as possible to the desired proportion. But that implementation also uses sorting. I would think there's a more efficient way to choose a random k-subset of n elements. I wonder if we could just use [numpy.random.choice](https://docs.scipy.org/doc/numpy/reference/generated/numpy.random.choice.html). Something like:
```
validation_session_ids = numpy.random.choice(num_sessions, fraction*num_sessions, replace=False)
# Use some numpy magic I don't know to convert to an array of length num_sessions with 1 in each row indicated by validation_session_ids
# Convert to filter SArray
```

igiloh replied:
Good idea - but it goes beyond the scope of this PR.  
Let's open a separate issue for that?

igiloh replied:
BTW - session_id may be a string - so using `random.seed(x) ; random.uniform(0,1)` as the "hash" function won't work. At least not out of the box without pre-processing (in the string case).  
I'll try to see if I can find some more suitable hash func.

nickjong commented on src/unity/python/turicreate/test/test_activity_classifier.py:
```diff
@@ -17,28 +17,59 @@
+    lines_in_each_session[-1] += num_examples - sum(lines_in_each_session)
+
+    if lines_in_each_session[-1] < 0:
+        raise ValueError()
```
This doesn't guard against `lines_in_each_session[-1] == 0`

And the probability of an error increases as the number of sessions increases.

It might be more robust just to normalize the sampled `lines_in_each_session`, by multiplying each value by `num_examples / sum(lines_in_each_session)`, followed by the existing fudge factor (in line 62) to deal with rounding.

(Interestingly, this sampling procedure would begin to resemble how you sample multinomial parameters from a Dirichlet distribution: https://en.wikipedia.org/wiki/Dirichlet_distribution#Random_number_generation)

igiloh replied:
I've known my way of guaranteeing `sum(lines_in_each_session) == num_examples` is off right when I've written it - but couldn't think of a better method. Yours is the better (and probably the correct) method. Thanks!

I'll read about the Dirichlet sampling tomorrow, when my mind is still fresh. Right now the very mention of it brings back haunting memories from horrible college courses... 😄

nickjong replied:
Eh, the link I posted is of mostly academic interest. The sampling procedure boils down to generating `num_sessions` draws from a random distribution and then normalizing to obtain a random list of `num_sessions` probabilities that sum to 1.

The Dirichlet connection comes in that you can view `lines_in_each_session` as a multinomial random variable that we could sample e.g. using https://en.wikipedia.org/wiki/Multinomial_distribution#Sampling_from_a_multinomial_distribution. But we don't really care about `lines_in_each_session` actually having a multinomial distribution. Sampling a Dirichlet is a convenient way of giving us something similar, literally choosing a random multinomial distribution (represented by its parameters).

Maybe the real question is to what degree the test even relies on `lines_in_each_session` being random....

igiloh replied:
None of our tests are truly random at the moment.  
They all set the seed to `42` (always!), and most don't even use randomness.  

Nevertheless - I do believe that we'll add proper pseudo-randomness to the tests someday - so it's better to write new tests with that mindset.

igiloh commented on src/unity/python/turicreate/toolkits/activity_classifier/util.py:
```diff
@@ -68,7 +70,43 @@ def random_split_by_session(dataset, session_id, fraction=0.9, seed=None):
+    # For a small number of unique sessions, creating a filter using SFrame.is_in() would
+    # be efficient - since filtering with a binary SArray maintains order, and no sorting
+    # is required.
+    if len(unique_sessions) < _MAX_SESSIONS_TO_USE_IS_IN:
```
For now - I've set some arbitrary hard limit, which I've found empirically.  
Giving it some more thought - the `is_in` solution takes o(k * n), where `k` is the number of sessions, and the filter+sort solution takes o(nlogn).   
Maybe I can improve this condition by setting `if number_of_sessions < log(len(dataset))`?  

@nickjong - what do you think? (that's all assuming I won't go with your suggested `apply` method).

nickjong replied:
I think what you have is fine. There's only so far we can get with asymptotic analysis, and if this continues to be a problem, I suspect it's a better investment of our time to find a way to make the O(n) approach work.

igiloh commented:
I've changed the `random_split_by_session()` implementation to the O(n) one suggested by @nickjong.  
The split accuracy doesn't seem be much affected - but the speedup is tremendous. Thanks Nick!

I've also changed the unit tests accordingly - there's no need to test a very large case, if the implementation is similar for all data sizes.

@nickjong - can you please give this PR a fresh review?

igiloh commented:
@nickjong - let's open a separate issue for your suggestion regarding `random_split` [improvements](https://github.com/apple/turicreate/pull/800#discussion_r199876248).

igiloh commented on src/unity/python/turicreate/toolkits/activity_classifier/util.py:
```diff
@@ -68,7 +70,28 @@ def random_split_by_session(dataset, session_id, fraction=0.9, seed=None):
+    def random_session_pick(session_id):
+        # If we will use only the session_id as the seed - the split will be constant for the
+        # same dataset across different runs, which is of course undesired
+        _random.seed(hash(session_id) + seed)
```
Sounds right. Will do.

igiloh commented:
@nickjong - I've handled the last two issues you've mentioned.  
Can you please look at the last commit quickly before I merge this PR?

Thanks!

nickjong commented:
Great, looks good to me! Let's go ahead and merge!
